jax/experimental/pjit.py

```diff
@@ -45,6 +46,99 @@ def pjit(fun: Callable,
          out_axis_resources,
          static_argnums: Union[int, Sequence[int]] = (),
          donate_argnums: Union[int, Sequence[int]] = ()):
+  """Makes ``fun`` compiled and automatically partitioned using XLA.
+
+  The returned function has semantics equivalent to those of ``fun``, but
+  is compiled to a single multi-device XLA computation. The partitioning
+  over multiple devices happens automatically, based on propagation of
+  input partitioning specified in ``in_axis_resources`` and output partitioning
+  specified in ``out_axis_resources``. The resources specified in those two
+  arguments can only refer to mesh axes, as defined by the
+  :py:func:`jax.experimental.maps.mesh` context manager. Note that the mesh
+  definition at ``pjit`` application time is ignored, and the returned function
+  will use the mesh definition available at each call site.
+
+  Inputs to ``pjit`` do not have to be partitioned accross multiple devices.
+  The implementation ensures to do the sharding automatically in that case.
+  However, for best performance it is best to ensure that the arguments
+  are correctly sharded (e.g. they are output of another ``pjit`` with
+  ``out_axis_resources`` matching the desired ``in_axis_resources``).
+
+  .. note::
+    Standard caveats apply to multi-process JAX invocations. ``pjit`` can operate
+    over meshes spanning multiple processes, but in that case the input dimensions
+    partitioned over multi-process mesh axes are expected to be of size matching
+    the size of the local mesh axis. The invocation of the ``pjit``ed function
+    will be then equivalent to running a computation with a global view of
+    the input array. In particular all elements coming from all processes are
+    visible to the computation. Outputs in each process will also correspond to
+    the slices of the global value that correspond to the data partitioned onto
+    the local devices.
```

Comment:
```suggestion
"""Makes ``fun`` compiled and automatically run across multiple devices.

  The returned function has semantics equivalent to those of ``fun``, but is
  compiled to an XLA computation that runs across multiple devices
  (e.g. multiple GPUs or multiple TPU cores). This can be useful if the jitted
  version of ``fun`` would not fit in a single device's memory, or to speed up
  ``fun`` by running each operation in parallel across multiple devices.

  Multi-device execution is achieved by spatially partitioning the data that
  flows through the computation, so each operation is run across all devices and
  each device runs only a piece of the full data. (Some data can optionally be
  replicated, which is sometimes more efficient for small arrays when combined
  with larger spatially-partitioned arrays.) Communication between devices is
  automatically inserted as necessary.

  The partitioning over devices happens automatically based on
  propagation of input partitioning specified in ``in_axis_resources`` and
  output partitioning specified in ``out_axis_resources``. The resources
  specified in those two arguments must refer to mesh axes, as defined by
  the :py:func:`jax.experimental.maps.mesh` context manager. Note that the mesh
  definition at ``pjit`` application time is ignored, and the returned function
  will use the mesh definition available at each call site.

.. note::

    **Multi-process platforms:** On multi-process platforms such as TPU pods,
    ``pjit`` can be used to run computations across all available devices across
    processes. To achieve this, ``pjit`` is designed to be used in SPMD Python
    programs, where every process is running the same Python code such that all
    processes run the same pjit'd function in the same order.

    When running in this configuration, the mesh should contain devices across
    all processes. However, any input argument dimensions partitioned over
    multi-process mesh axes should be of size equal to the corresponding *local*
    mesh axis size, and outputs will be similarly sized according to the local
    mesh. ``fun`` will still be executed across *all* devices in the mesh,
    including those from other processes, taking advantage of direct
    device-to-device communication. Conceptually, this can be thought of as
    running ``fun`` over arrays partitioned across processes, where each process
    "sees" only its local piece of the input and output, corresponding to its
    local sub-mesh.

    The SPMD model requires that the same multi-process pjit'd functions must be
    run in the same order on all processes, but they can be interspersed with
    arbitrary operations running in a single process.
```

Reply:
Hello, I would like to suggest overbearing editorial suggestions :P This is a combo of what you wrote + the `pmap` and `sharded_jit` docstrings. This is as far as I got and I'm OOO tomorrow and Friday, so let's continue next week if that's ok.

I completely omitted the paragraph about pre-sharding because it's slightly misleading: there should only be a performance advantage if you're passing the same input repeatedly, otherwise you can just let it do the sharding each time (I think?). I'm not opposed to mentioning this, but didn't have time to clarify it.

Reply:
The pre-sharding discussion is important (IMO) because it's the only place where we make clear that you should use the same axis resources/partitioning on both sides for an output that you want to feed in as an input.

Reply:
I also find the discussion of how the partitioning is achieved a little overly verbose for the docstring. The partitioning is just totally up to the compiler, and there's no need to mention that it might or might not choose to replicate parts of it. I also find the term "spatially partitioned" _extremely confusing_, so I'd rather avoid using it.

Reply:
Hi, sorry for the delay. I made a doc so we can iterate on this more easily: https://docs.google.com/document/d/1GDVicYyFvaKl0i_Bmggg_FDdUM4QQfZLTanIrIo-COA/edit (googlers only, if anyone else wants to contribute lemme know). I got rid of the "spatially partitioned" paragraphs and added a clarified version of the pre-sharding discussion, and made a few other tiny edits.